### DISCUSSION TITLE
Web notifications preferences for websites keep being erased

### DISCUSSION THREAD

**Discussion by BoffinBrain** [Cookie-AutoDelete/Cookie-AutoDelete#1432] [Q&A / Support] [answered]
For about a month, I've had intermittent issues with websites forgetting my preference to show web notifications. Some websites are in containers and some aren't. I suspect there has been a regression in CAD or something about Firefox's local storage may have changed. I'm aware that in CAD's container settings section, it shows the following warning:

> WARNING: Enabling Container Tabs with any of the additional browsing data cleanups above may cause unwanted side effects due to browser API limitations. Site data types, with the exception of cookies, will be cleared by hostname for ALL containers.

I'm guessing the issue might be related to this, but I must stress that the issue only started happening recently. What can I do to diagnose the problem further?

CAD version: 3.8.1
Firefox versions ~101 through to 104.0b2

**Comment by kennethtran93** [comment, marked as answer]
If you know now, there a 'debug mode' option in CAD settings. Follow those instructions to start looking at the console log lines (there will be alot). Clear the console logs, then execute the cleanup again. Save the output, and search/find the logs pertaining to the domain in question. They'll tell you what the extension has done (including checks to determine if it will be cleaned).

If you're ok with it, you can also share those logs to me and let me know which domain I should be looking for.

In debug mode, all cookie names are kept, but it's value is removed and replaced with `***` for display and export purposes only. The original cookie value is kept as is and not shown.

**Reply by BoffinBrain** [reply]
Thanks for the tip. I'll try and get something useful for you in the future, but first I have to find a reliable way to reproduce the issue. Simply closing other tabs and doing a clean doesn't cause the problem, and Firefox's little permissions popup next to the address bar continues to say that the website can send notifications. For now, I can paste my settings:

```
- activeMode: true
- delayBeforeClean: 600
- discardedCleanup: false
- domainChangeCleanup: true
- enableGreyListCleanup: true
- cleanCookiesFromOpenTabsOnStartup: false
- cleanExpiredCookies: true
- siteDataEmptyOnEnable: true
- cacheCleanup: true
- indexedDBCleanup: true
- localStorageCleanup: true
- pluginDataCleanup: true
- serviceWorkersCleanup: true
- contextualIdentities: true
- contextualIdentitiesAutoRemove: true
- statLogging: true
- showNumOfCookiesInIcon: true
- keepDefaultIcon: false
- showNotificationAfterCleanup: false
- manualNotifications: true
- notificationOnScreen: 3
- enableNewVersionPopup: false
- sizePopup: 16
- sizeSetting: 16
- contextMenus: false
- debugMode: true
```

**Reply by kennethtran93** [reply]
I'm not sure how they're doing web notifications - do they set a service worker?

If you are using containers/contextual identities, then ensure that the site in question is only accessed through that one container (There's a mozilla firefox addon that allows you to keep certain domains/sites to one container so you don't accidentally load it in another one).

**Reply by BoffinBrain** [reply]
I'm not up-to-speed with the latest web technologies so I'm not exactly sure how I'd go about checking whether they use service workers. The three sites I mostly want to focus on are Outlook, Twitter and Reddit.

Opening a website in a different container could indeed be part of the issue. What makes it trickier is that I have multiple Twitter accounts on multiple containers, so I'm not sure what may happen at that point if any of the containers, or the default one, does not have a whitelist rule set up.

**Reply by kennethtran93** [reply]
Yes in the event that the default container doesn't have twitter rule set up and it is accessed through that default container, then site data that is enabled for cleanup (which appears to be all of them) will be triggered.  Unfortunately this is not a CAD issue (mostly, as the ability to include containers as part of filter requirements weren't implemented earlier on), but the firefox webextension API issue.  While there is a select few that may be able to clean only those specified by containers, it is not all of them.

**Reply by BoffinBrain** [reply]
Okay, yes, that appears to be the case. I'll have to consider my settings a little more carefully. Thanks for the explanation!
